src/Compilers/Shared/DesktopBuildClient.cs

```diff
@@ -313,6 +317,55 @@ protected override int RunLocalCompilation(string[] arguments, BuildPaths buildP
             }
         }
 
+        // Protected for testing
+        protected static bool TryConnectToNamedPipeWithSpinWait(NamedPipeClientStream pipeStream,
+                                                                int timeoutMs,
+                                                                CancellationToken cancellationToken)
+        {
+            // .NET 4.5 implementation of NamedPipeStream.Connect busy waits the entire time.
+            // Work around is to spin wait.
+            const int maxWaitIntervalMs = 50;
+            int elapsedMs = 0;
+            var sw = new SpinWait();
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                int waitTime;
+                if (timeoutMs == Timeout.Infinite)
+                {
+                    waitTime = maxWaitIntervalMs;
+                }
+                else
+                {
+                    waitTime = Math.Min(timeoutMs - elapsedMs, maxWaitIntervalMs);
+                    if (waitTime <= 0)
+                    {
+                        return false;
```

Comment:
This timeout is even more constrained than I thought: it isn't set anywhere except in test code. In public endpoints it is always set to a constant non-negative value. I will add an assert that it is in the expected range.